File: erlang/euler10.erl
-module(euler10).
-export[euler10/0].

% Sum of all primes below 2000000
euler10() ->
    lists:foldl(
        fun(X, Sum) -> X + Sum end,
        0,
        primes(1999999)).

% Generate all primes up to (including) N
primes(N) when N < 2 -> []; 
primes(N) -> [2 | [X || X <- lists:seq(3, N, 2), isprime(X, 2)]].

% Check if Candidate is prime. Tail recursive "brute force" algorithm, must be started by setting Divisor=2
isprime(Candidate, Divisor) when Candidate rem Divisor == 0 -> false;
isprime(Candidate, Divisor) when Candidate < Divisor * Divisor -> true;
isprime(Candidate, Divisor) -> isprime(Candidate, Divisor + 1).

    
    

File: clojure/euler10.clj
; Sum of all primes below 2.000.000
(defn isprime
    ([n] (loop [x 2]
        (if (= 0 (mod n x))
	    false
	    (if (> (* x x) n)
	        true
		(recur (inc x)))))))

(defn primes
    ([] (cons 2 (filter (fn [x] (isprime x)) (iterate (fn [x] (+ x 2)) 3)))))

(println (
    reduce + (take-while (fn [x] (< x 2000000)) (primes))))

